### PR TITLE
Build in graphql support with automatic schema loading via Postgraphile. Also organize the configuration for easier understanding

### PR DESCRIPTION
**Pull request by dfellis**
Postgraphile was easier to set up than I expected, much faster to spin up than I expected, and we have much more control over it than Hasura. Still limiting things to just the local environment for now, but we can think about how to expose this in a safe, controlled way. (It *should* still require Auth0 auth in those environments with how I set it up, but that's not tested.)

I also made the config nicer (and documented) in the process because the big blob of a single JSON object is hard to follow and not necessary.

### CODE REVIEW

**Comment by dfellis** [comment]
Screenshot from 2022-04-15 10-39-53

**Comment by dfellis** [comment]
Method calling is via GraphQL `mutation` and it also works out of the box (and converts `snake_case` to `camelCase` so it looks normal to GraphQL users

Screenshot from 2022-04-15 11-07-58

File: src/config/staging.ts [hunk 1]
```diff
@@ -1,19 +1,28 @@
...
+    user: process.env.DB_USER ?? throwError('No DB User defined'),
+    password: process.env.DB_PASSWORD ?? throwError('No DB Password defined'),
+    port: 5432,
+    forceSSL: true,
+  },
+  logger: {
+    debug: true,
+    test: false,
+  },
+  auth0: {
```
Comment (depombo) [inline src/config/staging.ts:19]: maybe let's just do auth due to typos in the past between authO and auth0?

File: src/router/graphql.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      {
+        externalUrlBase: `/v1/graphql/${db}`,
+        graphiql: withGraphiql,
+      });
+    return postgraphiles[db];
+  };
+
+  graphql.use('/:db', async (req, res, next) => {
+    const db = config.auth0 ?
+      // This will throw an error if the user is not allowed to access the specified DB
```
Comment (depombo) [inline src/router/graphql.ts:28]: there is no try catch, does the error crash the process?

**Review by depombo — APPROVED** [review]

File: src/router/graphql.ts (hunk 2 above)
Comment (dfellis) [reply src/router/graphql.ts:28]: I assumed this would handle it? https://github.com/iasql/iasql-engine/blob/main/src/index.ts#L43-L52

It's a thrown error from an Express router and I thought Express grabbed the exceptions for us to route into that thing.

File: src/config/staging.ts (hunk 1 above)
Comment (dfellis) [reply src/config/staging.ts:19]: Sure

File: src/router/graphql.ts (hunk 2 above)
Comment (depombo) [reply src/router/graphql.ts:28]: Got it, I think we should be fine then!

**Review by depombo — APPROVED** [review]

**Review by depombo — APPROVED** [review]

File: src/config/config.ts [hunk 3]
```diff
@@ -1,17 +1,34 @@
...
+    test: boolean; // Whether or not a special test logger is enabled (bypass weirdness with Jest)
+  }
+  // Configuration for auth0 access control
+  auth?: { // Not including this sub-object implies it is not enabled
+    domain: string;
+    audience: string;
+  };
+  // Configuration for sentry error reporting
+  sentry?: { // Not including this sub-object implies it is not enabled
+    dsn: string;
```
Comment (aguillenv) [inline src/config/config.ts:26]: shouldn't this be `dns`? 😄

Comment (dfellis) [reply src/config/config.ts:26]: Nope. It's DSN. [Sentry Data Source Name](https://docs.sentry.io/product/sentry-basics/dsn-explainer/)
